### PR TITLE
[RELAY,TOPI] Threefry PRNG: splittable and stateless

### PR DESCRIPTION
**Pull request by tkonolige**
This PR adds a fast PRNG to Relay for use in dropout and batch norm. The PRNG is stateless: for a given key, it always returns the same random number. It is also splittable: for a given key, we can split the key to generate two new ones.

JAX provides a good explanation of stateless and splittable: https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#JAX-PRNG.

@altanh

### CODE REVIEW

**Comment by altanh** [comment]
awesome work, cc @tqchen @junrushao1994 @MarisaKirisame @eric-haibin-lin who may be interested

I think it may be worth discussing high vs low level API for this, and what namespace it should live in. I wrote a few examples for how we might use this here https://discuss.tvm.apache.org/t/rfc-handling-effect-in-tvm-and-relay/5946/25?u=altanh

**Comment by altanh** [comment]
worth noting that `threefry_generate` can only generate for shapes that have size as multiple of 4, can you update the examples and documentation in the Relay python op to reflect this?

**Comment by altanh** [comment]
## Naming
I propose we move everything PRNG to a new `random` namespace/module in Relay, so `relay.random.threefry_generate` etc.

## Handling different PRNG kernels

The splitting, keygen, and bit-gen operations will be kernel-specific. However, AFAIK, most if not all of the commonly used random ops simply require random bits as inputs (i.e. don't care about how those bits were generated). I propose the following approach to handling different kernels (thanks to @jroesch for helpful discussion):

- For each kernel `K` (e.g. `K = threefry`, `K = philox`, etc.), define `relay.random.K_key`, `relay.random.K_split` , and `relay.random.K_generate`.
- In Python (and any other host language where we plan on writing Relay code directly), define `relay.random.key(seed: int, K: RandomAlg)` (where `RandomAlg` is some kind of enum with members `RandomAlg.THREEFRY` etc.) and `relay.random.split(key, K: RandomAlg)`. We can set a default K (for now obviously Threefry) to make it easier to use.
- Now, for each random op `rop` that we care about, define `relay.random.rop_from_bits`, and in Python define `relay.random.rop(key, K)` which internally calls `relay.random.rop_from_bits(relay.random.K_generate(key))`. This should suffice to hide the algorithm from the user-facing API.

**Problems.**

- This might fail silently if the user inconsistently uses a key for kernel K with an op that uses kernel L ≠ K, if the keys are the same shape. This seems quite bad. Ideally, the keys will somehow be typed (perhaps we can use a Relay ADT?) but not sure how well supported this kind of use case is currently.

## Other notes
- I will open a follow-up PR that adds some initial operators (currently I have a uniform random op), along with support for non-multiple of 4 sizes.
- We should think about how to generalize over dtypes - currently, only `uint64` is supported for random bit generation in Threefry. Do we want to support more in the bit generation (is it possible?), or should we delegate this responsibility to auxiliary ops that split/join the `uint64`s as needed (e.g. converting 5 `uint64` to 10 `uint32`)? I think we need to come up with some kind of standardized approach for all the future random ops.

**Comment by tqchen** [comment]
cc @antinucleon

**Comment by altanh** [comment]
I've moved everything to a new `random` namespace (for Relay and TOPI) and fixed a small bug. This should be good to go I think.

File: src/relay/op/random/kernel.cc [hunk 1]
```diff
@@ -25,21 +25,23 @@ namespace relay {
 
 TVM_REGISTER_NODE_TYPE(ThreefryGenerateAttrs);
 
+static const TensorType THREEFRY_KEY_TYPE = TensorType({10}, tvm::DataType::UInt(64));
```
Comment (tqchen) [inline src/relay/op/random/kernel.cc:28]: As a rule of thumb, try to avoid static variables. Use static functions that returns these variables instead. As the construction is not that costly

Comment (altanh) [reply src/relay/op/random/kernel.cc:28]: Got it, is the naming style OK for a static function? `THREEFRY_KEY_TYPE()`?

Comment (tqchen) [reply src/relay/op/random/kernel.cc:28]: Usually CamelCase is preferred for a global or static function

**Comment by altanh** [comment]
bump @antinucleon @junrushao1994 @eric-haibin-lin @MarisaKirisame 

(please cc anyone else interested in PRNG for review, thanks!)

File: python/tvm/relay/op/random/kernel.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    This is useful if you have to subsequent calls which each need their own
+    independent random number generation.
+
+    Example
+    -------
+
+    .. code-block:: python
+
+        def foo(key):
+            new_key, num = threefry_generate(key, (1,))
```
Comment (altanh) [inline python/tvm/relay/op/random/kernel.py:107]: fix this example to use multiple of 4 for now, until we support non-multiples

File: python/tvm/topi/random/kernel.py [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+                        _ROTATIONS[nwords][(i * 4 + k) % 8][j],
+                    )
+                for j in range(nwords):
+                    tmp[tmp_offset + l * nwords + j] = out_buf[
+                        out_offset + l * nwords + _PERMUTATIONS[nwords][j]
+                    ]
+                # number of rounds is even, so out always contains the result
+                (out_buf, tmp) = (tmp, out_buf)
+                (out_offset, tmp_offset) = (tmp_offset, out_offset)
+
```
Comment (altanh) [inline python/tvm/topi/random/kernel.py:185]: I see some `TODO` in this function (`_threefry`), do they affect the correctness of the algorithm?

File: python/tvm/topi/random/kernel.py [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+        [out_gen.shape, out_array.shape],
+        [gen],
+        lambda ins, outs: gen_ir(ins[0], outs[0], outs[1]),
+        out_buffers=[out_gen, out_array],
+        name="threefry_generate",
+        tag="threefry_generate",
+    )
+
+
+def _shift_right(irb, a, b, out_a, a_off, out_b, b_off):
```
Comment (altanh) [inline python/tvm/topi/random/kernel.py:304]: for clarity maybe comment if this is logical or arithmetic right shift

**Review by altanh — COMMENTED** [review]
@tkonolige I had some additional questions about TODOs in the Threefry kernel, maybe you can clarify?

File: python/tvm/topi/random/kernel.py (hunk 3 above)
Comment (tkonolige) [reply python/tvm/topi/random/kernel.py:186]: The only one that matters is `TODO should be wrapping`. I do not know if TVM guarantees unsigned integer arithmetic to be wrapping (instead of saturating).

**Comment by tmoreau89** [comment]
Thanks for the reviews, @junrushao1994 @MarisaKirisame @eric-haibin-lin if you have some cycles, input would be appreciated!

**Review by jwfromm — APPROVED** [review]
Really well written and commented code, great work!

File: src/relay/op/random/kernel.cc [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+#include <tvm/relay/attrs/random.h>
+#include <tvm/relay/op.h>
+
+namespace tvm {
+namespace relay {
+
+TVM_REGISTER_NODE_TYPE(ThreefryGenerateAttrs);
+
+static TensorType ThreefryKeyType() { return TensorType({10}, tvm::DataType::UInt(64)); }
```
Comment (MarisaKirisame) [inline src/relay/op/random/kernel.cc:28]: sorry for not pointing this out earlier. Maybe a good thing to do is to wrap this in a newtype? (e.g. define a type ThreefryKey that you cannot use in anyway except in random operation. this will avoid doing arithmetic on random seed.)

**Comment by tqchen** [comment]
also cc @yzhliu @hzfan @comaniac who might be interested

File: tests/python/relay/test_prng.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+def test_threefry_repeatability(target, ctx):
+    target, ctx = "llvm", tvm.cpu(0)
+    key1 = tvm.relay.random.threefry_key(1)
+    rand1 = tvm.relay.random.threefry_generate(key1, (12,))
+    out_key1, out1 = tvm.relay.create_executor(
+        "vm", tvm.IRModule.from_expr(tvm.relay.Function([], rand1)), target=target, ctx=ctx
+    ).evaluate()()
+
+    key2 = tvm.relay.random.threefry_key(1)
+    rand2 = tvm.relay.random.threefry_generate(key2, (12,))
```
Comment (electriclilies) [inline tests/python/relay/test_prng.py:34]: It might be a good idea to add some tests with shapes other than (12, ), especially in the type test.

**Review by electriclilies — APPROVED** [review]
Overall this looks good to me! I just have a few design questions and notes on comments and your tests.

File: src/relay/op/random/kernel.cc (hunk 5 above)
Comment (tkonolige) [reply src/relay/op/random/kernel.cc:28]: Adding a new opaque new type to tvm seems really involved. We have to add a new visitor for each type visitor, which seems like it may cause issues with some passes. We'd also have to add a no-op function with implementations to satisfy the type checker. Or we'd have to add a wrapper struct with all the proper conversion functions. Given all this complication, I don't think it is a good idea.

File: python/tvm/relay/op/random/kernel.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    ----------
+    key : relay.Expr
+        key that uniquely determines the random values. Multiple uses with the
+        same key will generate the same random values. This key should be
+        treated as an opaque pointer. You can create one from calling
+        :py:func:`threefry_key`, :py:func:`threefry_split`, or
+        :py:func:`threefry_generate`. **Do not use this key again after calling
+        this function.**
+
+    shape : Sequence[int]
```
[the remark this answers is not in the document]
Comment (tkonolige) [reply python/tvm/relay/op/random/kernel.py:79]: It is an implementation detail. Basically, threefry uses 4 64-bit words as its state, inputs, and outputs.

File: python/tvm/relay/op/random/kernel.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    random_array : relay.Expr
+        Array of random numbers. Has shape `shape`.
+    """
+    return _make.threefry_generate(key, shape)
+
+
+def threefry_split(key):
+    """Split an existing Threefry key into two new ones.
+
+    This is useful if you have to subsequent calls which each need their own
```
[the remark this answers is not in the document]
Comment (tkonolige) [reply python/tvm/relay/op/random/kernel.py:98]: Creating separate keys has not been theoretically proven to be as random as splitting a single key. Maybe I should add a comment that you should only really create one key. On the other hand, these details might be better handled at a higher level interface (future work).

File: python/tvm/topi/random/kernel.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+#   Hashing". Proceedings of Haskell Symposium 2013 pp. 47-58.  MLA
+# - Ferguson, Niels, et al. "The Skein hash function family." Submission to NIST (round 3) 7.7.5
+#   (2010): 3.
+
+
+# Threefry is a counter based PRNG: given a unique input, it generates a unique random number. As
+# there is no state to maintain, we can apply it to a sequence of numbers (0..N) to generate a
+# sequence of random numbers in parallel. In order to make the PRNG splittable (that is we can
+# generate a sequence of random numbers in one place, and another sequence in another), we add a
+# path and key in addition to the counter. The path allows us to encode a sequence of splits (a 0 in
```
[the remark this answers is not in the document]
Comment (tkonolige) [reply python/tvm/topi/random/kernel.py:38]: The last sentence explains what the key is: "To avoid continuously growing the path, we can compress an existing path into the key portion of the generator by hashing the current key, path, and counter to create the new key (this same technique is used if we run out of room for the counter)." I've added a comment on how it is initialized.

I've also added an explanation of how random numbers are generated (we apply the hash to key, path, and counter).

**Review by MarisaKirisame — APPROVED** [review]

**Comment by tmoreau89** [comment]
Thank you @altanh @electriclilies @tqchen @jwfromm @MarisaKirisame for the reviews, the PR has been merged.

**Comment by yzhliu** [comment]
nice work. do we have plan to support cuda? Looks like it works on cpu only at this moment.

**Comment by tkonolige** [comment]
It should be easy to use on the GPU, just parallelize the outer loop.
